ariadne_codegen/plugins/shorter_results.py

```diff
@@ -0,0 +1,283 @@
+
+            # The module we import from is always the same as the method we
+            # modified when changing the return type.
+            if stmt.module not in self.extended_imports:
+                continue
+
+            # Add all additional imports discovered when generating the client
+            # method.
+            for additional_import in self.extended_imports[stmt.module]:
+                stmt.names.append(ast.alias(name=additional_import))
+
+        return super().generate_client_module(module)
+
+    def generate_client_method(
+        self, method_def: Union[ast.FunctionDef, ast.AsyncFunctionDef]
+    ) -> Union[ast.FunctionDef, ast.AsyncFunctionDef]:
+        """
+        Change the method generated in the client to call the inner field and
+        change the return type.
+        """
+        if len(method_def.body) < 1:
+            return super().generate_client_method(method_def)
+
+        last_stmt = method_def.body[-1]
+
+        if isinstance(last_stmt, ast.Return):
+            return self._generate_sync_client_method(method_def, last_stmt)
+        elif isinstance(last_stmt, ast.AsyncFor):
+            return self._generate_async_client_method(method_def, last_stmt)
+
+        return super().generate_client_method(method_def)
+
+    def _generate_async_client_method(
+        self,
+        method_def: Union[ast.FunctionDef, ast.AsyncFunctionDef],
+        async_for_stmt: ast.AsyncFor,
+    ):
+        if not isinstance(method_def.returns, ast.Subscript):
+            return super().generate_client_method(method_def)
+
+        if not isinstance(method_def.returns.slice, ast.Name):
+            return super().generate_client_method(method_def)
+
+        node_and_class = _return_or_yield_node_and_class(
+            method_def.returns.slice.id, self.class_dict
+        )
+        if node_and_class is None:
+            return super().generate_client_method(method_def)
+
+        yield_node, single_field_class, single_field_return_class = node_and_class
+
+        # Update the return type to be the inner field and ensure we reference
+        # the single field before returning.
+        method_def.returns = generate_subscript(
+            value=generate_name("AsyncIterator"),
+            slice_=yield_node,
+        )
+
+        if not isinstance(method_def.body[-1], ast.AsyncFor):
+            return super().generate_client_method(method_def)
+
+        if len(method_def.body[-1].body) < 1:
+            return super().generate_client_method(method_def)
+
+        if not isinstance(method_def.body[-1].body[0], ast.Expr):
+            return super().generate_client_method(method_def)
+
+        if not isinstance(method_def.body[-1].body[0].value, ast.Yield):
+            return super().generate_client_method(method_def)
+
+        previous_yield_value = method_def.body[-1].body[0].value.value
+        if previous_yield_value is None:
+            return super().generate_client_method(method_def)
```

Comment:
This is to make the type checker happy but it feels s bit verbose. I used similar patterns in some other places. 

Is this idiomatic enough? Can I simplify this when working with multiple `ast.stmt`?

Reply:
Split this to a separate method